godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Basis.kt

```diff
@@ -0,0 +1,616 @@
+        yx: Number,
+        yy: Number,
+        yz: Number,
+        zx: Number,
+        zy: Number,
+        zz: Number
+    ) :
+        this(Vector3(xx, xy, xz), Vector3(yx, yy, yz), Vector3(zx, zy, zz))
+
+
+    constructor(from: Vector3) : this() {
+        setEuler(from)
+    }
+
+    constructor(quat: Quat) : this() {
+        val d = quat.lengthSquared()
+        val s = 2.0 / d
+        val xs = quat.x * s
+        val ys = quat.y * s
+        val zs = quat.z * s
+        val wx = quat.w * xs
+        val wy = quat.w * ys
+        val wz = quat.w * zs
+        val xx = quat.x * xs
+        val xy = quat.x * ys
+        val xz = quat.x * zs
+        val yy = quat.y * ys
+        val yz = quat.y * zs
+        val zz = quat.z * zs
+        set(
+            1.0 - (yy + zz), xy - wz, xz + wy,
+            xy + wz, 1.0 - (xx + zz), yz - wx,
+            xz - wy, yz + wx, 1.0 - (xx + yy)
+        )
+    }
+
+    constructor(axis: Vector3, phi: Double) : this() {
+        // Rotation matrix from axis and angle, see https://en.wikipedia.org/wiki/Rotation_matrix#Rotation_matrix_from_axis_and_angle
+        val axisq = Vector3(axis.x * axis.x, axis.y * axis.y, axis.z * axis.z)
+
+        val cosine: Double = cos(phi)
+        val sine: Double = sin(phi)
+
+        apply {
+            x.x = axisq.x + cosine * (1.0 - axisq.x)
+            x.y = axis.x * axis.y * (1.0 - cosine) - axis.z * sine
+            x.z = axis.z * axis.x * (1.0 - cosine) + axis.y * sine
+
+            y.x = axis.x * axis.y * (1.0 - cosine) + axis.z * sine
+            y.y = axisq.y + cosine * (1.0 - axisq.y)
+            y.z = axis.y * axis.z * (1.0 - cosine) - axis.x * sine
+
+            z.x = axis.z * axis.x * (1.0 - cosine) - axis.y * sine
+            z.y = axis.y * axis.z * (1.0 - cosine) + axis.x * sine
+            z.z = axisq.z + cosine * (1.0 - axisq.z)
+        }
+    }
+
+    internal constructor(native: CValue<godot_basis>) : this() {
+        memScoped {
+            this@Basis.setRawMemory(native.ptr)
+        }
+    }
+
+    internal constructor(mem: COpaquePointer) : this() {
+        this.setRawMemory(mem)
+    }
+
+
+    //INTEROP
+    override fun getRawMemory(memScope: MemScope): COpaquePointer {
+        val value = cValue<godot_basis_layout> {
+            x.x = this@Basis.x.x.toFloat()
+            x.y = this@Basis.x.y.toFloat()
+            x.z = this@Basis.x.z.toFloat()
+            y.x = this@Basis.y.x.toFloat()
+            y.y = this@Basis.y.y.toFloat()
+            y.z = this@Basis.y.z.toFloat()
+            z.x = this@Basis.z.x.toFloat()
+            z.y = this@Basis.z.y.toFloat()
+            z.z = this@Basis.z.z.toFloat()
+        }
+        return value.getPointer(memScope)
+    }
+
+    override fun setRawMemory(mem: COpaquePointer) {
+        val value = mem.reinterpret<godot_basis_layout>().pointed
+        x.setRawMemory(value.x.ptr)
+        y.setRawMemory(value.y.ptr)
+        z.setRawMemory(value.z.ptr)
+    }
+
+
+    //API
```

Comment:
What do you mean by `API`?
Edit: I suppose you used the comments from the godot source code; maybe make it a bit clearer or just remove the comments. IMHO the visibility should define the "api" functions

Reply:
It's just me who structured the code that way:
Constant
Constructors
API
Utilities

Not all public method are API functions. That's why I got a separate Utilities part for operator for example. 
I can remove them if you want. That's just how I do things.

Reply:
No you don't have to, I'm fine with it if you prefer it that way.